In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
data = pd.read_csv('E:/Data Science AI Ecosystem/DATASETS/DATASETS 2/exam.csv')
data.head()

,5me,guide,result
0,0,0,fail
1,1,1,pass
2,1,0,pass
3,0,0,fail
4,0,0,fail


In [3]:
data

,5me,guide,result
0,0,0,fail
1,1,1,pass
2,1,0,pass
3,0,0,fail
4,0,0,fail
5,1,1,pass
6,0,1,fail
7,1,1,pass
8,1,0,fail
9,1,1,pass


In [4]:
data.result.value_counts()

result
fail    5
pass    5
Name: count, dtype: int64

- Parent Node Entropy -

In [5]:
entropy_results = -(5/10)*np.log2(5/10) - (5/10)*np.log2(5/10)
entropy_results

1.0

- Now let's calculate the entropy of each feature and check whose has more information gain. The feature which has more information gain the data will be split according to that feature.

In [6]:
data.rename(columns = {'5me':'yt'}, inplace = True)

In [7]:
data

,yt,guide,result
0,0,0,fail
1,1,1,pass
2,1,0,pass
3,0,0,fail
4,0,0,fail
5,1,1,pass
6,0,1,fail
7,1,1,pass
8,1,0,fail
9,1,1,pass


In [8]:
data[data.yt == 1]

,yt,guide,result
1,1,1,pass
2,1,0,pass
5,1,1,pass
7,1,1,pass
8,1,0,fail
9,1,1,pass


- Calculating the Entropy of Yt with yes samples -

In [9]:
entropy_results_yt_yes = -(5/6)*np.log2(5/6) -(1/6)*np.log2(1/6)
entropy_results_yt_yes

0.6500224216483541

In [10]:
data[data.yt == 0]

,yt,guide,result
0,0,0,fail
3,0,0,fail
4,0,0,fail
6,0,1,fail


- Calculating the Entropy of Yt feature with No samples -

In [ ]:
# Zero entropy because all the values are same.

- Calculating the Combined Entropy of Yt with Yes and No samples entropy - 

In [11]:
yt_entropy = (6/10) * entropy_results_yt_yes +  (4/10) * 0
yt_entropy

0.39001345298901247

In [12]:
info_yt = entropy_results - (6/10) * entropy_results_yt_yes -  (4/10) * 0
info_yt

0.6099865470109875

The Information Gain is 0.6099

- Calculating the Entropy of Guide feature with Yes samples -

In [13]:
data[data.guide == 1]

,yt,guide,result
1,1,1,pass
5,1,1,pass
6,0,1,fail
7,1,1,pass
9,1,1,pass


In [14]:
entropy_results_guide_yes = -(4/5)*np.log2(4/5) -(1/5)*np.log2(1/5)
entropy_results_guide_yes

0.7219280948873623

- Calculating the Entropy of Guide feature with No samples -

In [15]:
data[data.guide == 0]

,yt,guide,result
0,0,0,fail
2,1,0,pass
3,0,0,fail
4,0,0,fail
8,1,0,fail


In [16]:
entropy_results_guide_no = -(1/5)*np.log2(1/5) -(4/5)*np.log2(4/5)
entropy_results_guide_no

0.7219280948873623

- Combined Entropy of Guide feature with Yes and No data samples -

In [17]:
guide_entropy = (5/10) * entropy_results_guide_yes +  (5/10) * entropy_results_guide_no
guide_entropy

0.7219280948873623

In [18]:
info_guide = entropy_results - guide_entropy
info_guide

0.2780719051126377

- The information Gain from guide feature is 0.278
- The info gain(yt) > info gain(guide)
- Hence the splitting will happen according to yt feature

### Implementing the Decision Tree Classifier -

In [19]:
from sklearn.preprocessing import LabelEncoder

In [20]:
le = LabelEncoder()
df = data
df.result = le.fit_transform(df.result)
df

,yt,guide,result
0,0,0,0
1,1,1,1
2,1,0,1
3,0,0,0
4,0,0,0
5,1,1,1
6,0,1,0
7,1,1,1
8,1,0,0
9,1,1,1


In [21]:
x = df[['yt','guide']]
y = df['result']

In [22]:
from sklearn.tree import DecisionTreeClassifier, export_graphviz

In [23]:
dt = DecisionTreeClassifier(criterion='entropy')

In [24]:
dt.fit(x,y)

DecisionTreeClassifier(criterion='entropy')

In [25]:
export_graphviz(dt,'dt.tree',feature_names=['yt','guide'])

In [26]:
dt.predict([[1,1]])

c:\Users\user\miniconda3\lib\site-packages\sklearn\base.py:465: UserWarning: X does not have valid feature names, but DecisionTreeClassifier was fitted with feature names
  warnings.warn(


array([1])

In [27]:
dt.score(x,y)

0.9